zebra-state/src/config.rs

```diff
@@ -1,4 +1,7 @@
-use std::path::PathBuf;
+use std::{
+    fs::{remove_dir_all, DirEntry, ReadDir},
+    path::PathBuf,
+};
 
 use serde::{Deserialize, Serialize};
```

Comment:
In async functions, we should use async filesystem operations.

`std` filesystem operations will block all the async tasks on the same thread, which is what we're trying to avoid.

```suggestion
use std::path::PathBuf;

use serde::{Deserialize, Serialize};
tokio::fs::{remove_dir_all, DirEntry, ReadDir};
```

Reply:
If we do this change there will be a lot of additional work to be done:

```
oxarbitrage@oxarbitrage-Lenovo-ideapad-320S-14IKB:~/zebra/issue1213/zebra$ cargo clippy --all-features --all-targets
   Compiling zebrad v1.0.0-beta.11 (/home/oxarbitrage/zebra/issue1213/zebra/zebrad)
    Checking zebra-state v1.0.0-beta.11 (/home/oxarbitrage/zebra/issue1213/zebra/zebra-state)
error[E0599]: the method `flatten` exists for struct `tokio::fs::ReadDir`, but its trait bounds were not satisfied
   --> zebra-state/src/config.rs:140:31
    |
140 |         for entry in read_dir.flatten() {
    |                               ^^^^^^^ method cannot be called on `tokio::fs::ReadDir` due to unsatisfied trait bounds
    |
   ::: /home/oxarbitrage/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.2/src/fs/read_dir.rs:57:1
    |
57  | pub struct ReadDir(State);
    | -------------------------- doesn't satisfy `tokio::fs::ReadDir: std::iter::Iterator`
    |
note: the following trait bounds were not satisfied because of the requirements of the implementation of `std::iter::Iterator` for `_`:
      `tokio::fs::ReadDir: std::iter::Iterator`
   --> zebra-state/src/service/block_iter.rs:70:6
    |
70  | impl Iterator for Iter<'_> {
    |      ^^^^^^^^     ^^^^^^^^

error[E0599]: no method named `is_ok` found for opaque type `impl futures::Future<Output = std::result::Result<(), std::io::Error>>` in the current scope
   --> zebra-state/src/config.rs:145:64
    |
145 |                         if remove_dir_all(delete_path.clone()).is_ok() {
    |                                                                ^^^^^ method not found in `impl futures::Future<Output = std::result::Result<(), std::io::Error>>`
    |
help: consider `await`ing on the `Future` and calling the method on its `Output`
    |
145 |                         if remove_dir_all(delete_path.clone()).await.is_ok() {
    |                                                                ++++++

error[E0308]: mismatched types
   --> zebra-state/src/config.rs:158:25
    |
158 |             return Some(read_dir);
    |                         ^^^^^^^^ expected struct `tokio::fs::ReadDir`, found struct `std::fs::ReadDir`

error[E0308]: mismatched types
   --> zebra-state/src/config.rs:165:12
    |
165 |     if let Ok(file_type) = entry.file_type() {
    |            ^^^^^^^^^^^^^ expected opaque type, found enum `std::result::Result`
    |
note: while checking the return type of the `async fn`
   --> /home/oxarbitrage/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.2/src/fs/read_dir.rs:285:38
    |
285 |     pub async fn file_type(&self) -> io::Result<FileType> {
    |                                      ^^^^^^^^^^^^^^^^^^^^ checked the `Output` of this `async fn`, expected opaque type
    = note: expected opaque type `impl futures::Future<Output = std::result::Result<std::fs::FileType, std::io::Error>>`
                      found enum `std::result::Result<_, _>`
help: consider `await`ing on the `Future`
    |
165 |     if let Ok(file_type) = entry.file_type().await {
    |                                             ++++++

Some errors have detailed explanations: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `zebra-state` due to 4 previous errors
warning: build failed, waiting for other jobs to finish...
^C  Building [=====================> ] 1193/1214: zebra-state(test)                                                                                                    
oxarbitrage@oxarbitrage-Lenovo-ideapad-320S-14IKB:~/zebra/issue1213/zebra$ 
```

Reply:
I tried making some of those changes, but it seems like `spawn_blocking` is easier.

So I just launched the whole delete task in a separate thread in 670533f38719a51d351d907a5101e03745dfd2a4.

Reply:
I think it will be nice to test some suggestions before making them. I understand that this is a lot more work than just adding a suggestion of code change without seeing anything else but i think it worth it sometimes. This is one of the things that @jvff used to do. If you realize the suggestion is too complicated then you don't even add it, if it is how you expected then we have a suggestion that can be implemented with the click of a button.

Reply:
Thanks for that feedback, I will try to only make suggestions that I am sure will work.